rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,311 @@
+
+const PARTITION_KEY_NAME: &str = "key";
+const PARTITION_KEY_VALUE: &str = "delta-rs";
+const OWNER_NAME: &str = "ownerName";
+const RECORD_VERSION_NUMBER: &str = "recordVersionNumber";
+const IS_RELEASED: &str = "isReleased";
+const LEASE_DURATION: &str = "leaseDuration";
+
+mod expressions {
+    pub const ACQUIRE_LOCK_THAT_DOESNT_EXIST: &str = "attribute_not_exists(#pk)";
+
+    pub const PK_EXISTS_AND_IS_RELEASED: &str = "attribute_exists(#pk) AND #ir = :ir";
+
+    pub const PK_EXISTS_AND_RVN_IS_THE_SAME: &str = "attribute_exists(#pk) AND #rvn = :rvn";
+}
+
+mod vars {
+    pub const PK_PATH: &str = "#pk";
+    pub const RVN_PATH: &str = "#rvn";
+    pub const RVN_VALUE: &str = ":rvn";
+    pub const IS_RELEASED_PATH: &str = "#ir";
+    pub const IS_RELEASED_VALUE: &str = ":ir";
+}
+
+pub struct DynamoLock {
+    client: DynamoDbClient,
+    table_name: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct LockItem {
+    owner_name: String,
+    record_version_number: String,
+    lease_duration: u64,
+    is_released: bool,
+
+    lookup_time: u128,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum DynamoError {
+    #[error("Dynamo table not found")]
+    TableNotFound,
+
+    #[error("Condition failed: {0}")]
+    ConditionalCheckFailed(String),
+
+    #[error("DynamoDB item has invalid schema")]
+    InvalidItemSchema,
+
+    #[error("Could not acquire lock for {0} sec")]
+    TimedOut(u64),
+
+    #[error("Get item error: {0}")]
+    GetItemError(RusotoError<GetItemError>),
+
+    #[error("Put item error: {0}")]
+    PutItemError(RusotoError<PutItemError>),
+}
+
+impl DynamoLock {
+    pub fn new() -> Self {
+        let client = DynamoDbClient::new(Region::default());
+        Self {
+            client,
+            table_name: String::from("mykhailo_test_table_2"),
+        }
+    }
+
+    async fn acquire_lock(&self) -> Result<LockItem, DynamoError> {
+        let mut lock_trying_to_be_acquired: Option<LockItem> = None;
+        let started = Instant::now();
+        let mut dur_to_wait = Duration::from_secs(1);
+
+        println!("Start {:?}", started);
+        while started.elapsed() < dur_to_wait {
+            let existing_lock = self.get_lock_item().await?;
+            let new_lock = LockItem::new();
+
+            let call = match existing_lock {
+                None => {
+                    println!("upsert_new_lock");
+                    self.upsert_new_lock(&new_lock).await
+                }
+                Some(ref l) if l.is_released => {
+                    println!("upsert_released_lock");
+                    self.upsert_released_lock(&new_lock).await
+                }
+                Some(ref existing) => match lock_trying_to_be_acquired {
+                    None => {
+                        println!("lock_trying_to_be_acquired None");
+                        lock_trying_to_be_acquired = existing_lock.clone();
+                        dur_to_wait = Duration::from_secs(5 + existing.lease_duration);
```

Comment:
The value you're setting `dur_to_wait` to here smells funny to me. From [LockItem::new](https://github.com/delta-io/delta-rs/pull/153/files#diff-b4ae09784de4beabd0a09649f76c108935196dfde86adf592a0b9bd95832eed9R261), I'm expecting lease duration to be the number of seconds to acquire the lock for. 

Would something like `now - (item.lookup_time + item.lease_duration)` be a good wait time?